File: test-py/test_path_changes.py
import json
import pathlib

import pytest

from fontra.core.changes import applyChange
from fontra.core.path import PackedPath

testDataPath = (
    pathlib.Path(__file__).parent.parent / "test-common" / "path-change-test-data.json"
)

pathChangeTestData = json.loads(testDataPath.read_text(encoding="utf-8"))
pathChangeTestInputData = pathChangeTestData["inputPaths"]


pathChangeTestData = [
    (
        testCase["testName"],
        testCase["inputPathName"],
        testCase["change"],
        PackedPath.fromUnpackedContours(testCase["expectedPath"]),
    )
    for testCase in pathChangeTestData["tests"]
]


@pytest.mark.parametrize(
    "testName, inputPathName, change, expectedData", pathChangeTestData
)
def test_applyChange(testName, inputPathName, change, expectedData):
    subject = PackedPath.fromUnpackedContours(pathChangeTestInputData[inputPathName])
    applyChange(subject, change)
    assert subject == expectedData

File: src-js/fontra-core/tests/test-path-changes.js
import { expect } from "chai";
import fs from "fs";

import { applyChange } from "@fontra/core/changes.js";
import { VarPackedPath } from "@fontra/core/var-path.js";

import { dirname, join } from "path";
import { fileURLToPath } from "url";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("Path Changes Tests", () => {
  const test_data_path = join(
    dirname(__dirname),
    "..",
    "..",
    "test-common",
    "path-change-test-data.json"
  );
  const test_data = JSON.parse(fs.readFileSync(test_data_path, "utf8"));
  const inputPaths = test_data["inputPaths"];
  const tests = test_data["tests"];

  for (let i = 0; i < tests.length; i++) {
    const test = tests[i];
    const testName = test["testName"];
    const inputPathName = test["inputPathName"];
    const expectedPath = VarPackedPath.fromUnpackedContours(test["expectedPath"]);

    const subject = VarPackedPath.fromUnpackedContours(
      copyObject(inputPaths[inputPathName])
    );
    it(`Apply Path Changes test #${i} -- ${testName}`, () => {
      applyChange(subject, test["change"]);
      expect(subject).to.deep.equal(expectedPath);
    });
  }
});

function copyObject(obj) {
  return JSON.parse(JSON.stringify(obj));
}
